File: ios/MiszIQTests/SettingsTests.swift
import XCTest
@testable import MiszIQ

/// Unit tests for settings system, theme modes, and audio/haptic managers
final class SettingsTests: XCTestCase {

    // MARK: - ThemeMode Tests

    func testThemeMode_system_shouldReturnNilColorScheme() {
        let themeMode = ThemeMode.system
        XCTAssertNil(themeMode.colorScheme, "System theme mode should return nil colorScheme")
    }

    func testThemeMode_light_shouldReturnLightColorScheme() {
        let themeMode = ThemeMode.light
        XCTAssertEqual(themeMode.colorScheme, .light, "Light theme mode should return .light colorScheme")
    }

    func testThemeMode_dark_shouldReturnDarkColorScheme() {
        let themeMode = ThemeMode.dark
        XCTAssertEqual(themeMode.colorScheme, .dark, "Dark theme mode should return .dark colorScheme")
    }

    func testThemeMode_allCases_shouldHaveDisplayName() {
        for mode in ThemeMode.allCases {
            XCTAssertFalse(mode.displayName.isEmpty, "\(mode) should have a display name")
        }
    }

    func testThemeMode_allCases_shouldHaveIcon() {
        for mode in ThemeMode.allCases {
            XCTAssertFalse(mode.icon.isEmpty, "\(mode) should have an icon")
        }
    }

    func testThemeMode_count_shouldBe3() {
        XCTAssertEqual(ThemeMode.allCases.count, 3, "There should be exactly 3 theme modes")
    }

    func testThemeMode_rawValues_shouldBeCorrect() {
        XCTAssertEqual(ThemeMode.system.rawValue, "system")
        XCTAssertEqual(ThemeMode.light.rawValue, "light")
        XCTAssertEqual(ThemeMode.dark.rawValue, "dark")
    }

    func testThemeMode_initFromRawValue_shouldWork() {
        XCTAssertEqual(ThemeMode(rawValue: "system"), .system)
        XCTAssertEqual(ThemeMode(rawValue: "light"), .light)
        XCTAssertEqual(ThemeMode(rawValue: "dark"), .dark)
        XCTAssertNil(ThemeMode(rawValue: "invalid"))
    }

    // MARK: - SoundEffect Tests

    func testSoundEffect_allCases_shouldHaveFileName() {
        for effect in SoundEffect.allCases {
            XCTAssertFalse(effect.rawValue.isEmpty, "\(effect) should have a file name")
        }
    }

    func testSoundEffect_count_shouldBe4() {
        XCTAssertEqual(SoundEffect.allCases.count, 4, "There should be exactly 4 sound effects")
    }

    func testSoundEffect_rawValues_shouldBeCorrect() {
        XCTAssertEqual(SoundEffect.correctAnswer.rawValue, "correct_chime")
        XCTAssertEqual(SoundEffect.wrongAnswer.rawValue, "wrong_buzz")
        XCTAssertEqual(SoundEffect.buttonTap.rawValue, "button_tap")
        XCTAssertEqual(SoundEffect.gameComplete.rawValue, "celebration")
    }

    // MARK: - Settings Default Values Tests

    func testSettings_musicEnabled_defaultShouldBeTrue() {
        // Default value should be true (music enabled by default)
        let defaultValue = true
        XCTAssertTrue(defaultValue, "Music should be enabled by default")
    }

    func testSettings_soundEffectsEnabled_defaultShouldBeTrue() {
        // Default value should be true (sound effects enabled by default)
        let defaultValue = true
        XCTAssertTrue(defaultValue, "Sound effects should be enabled by default")
    }

    func testSettings_hapticFeedbackEnabled_defaultShouldBeTrue() {
        // Default value should be true (haptic feedback enabled by default)
        let defaultValue = true
        XCTAssertTrue(defaultValue, "Haptic feedback should be enabled by default")
    }

    func testSettings_themeMode_defaultShouldBeSystem() {
        // Default theme mode should be system
        let defaultMode = ThemeMode.system
        XCTAssertEqual(defaultMode, .system, "Default theme mode should be system")
    }

    // MARK: - Settings Toggle Tests

    func testSettings_toggleMusic_shouldInvertValue() {
        var musicEnabled = true
        musicEnabled.toggle()
        XCTAssertFalse(musicEnabled, "Toggling music should invert the value")
        musicEnabled.toggle()
        XCTAssertTrue(musicEnabled, "Toggling again should restore the value")
    }

    func testSettings_toggleSoundEffects_shouldInvertValue() {
        var soundEnabled = true
        soundEnabled.toggle()
        XCTAssertFalse(soundEnabled, "Toggling sound effects should invert the value")
        soundEnabled.toggle()
        XCTAssertTrue(soundEnabled, "Toggling again should restore the value")
    }

    func testSettings_toggleHapticFeedback_shouldInvertValue() {
        var hapticEnabled = true
        hapticEnabled.toggle()
        XCTAssertFalse(hapticEnabled, "Toggling haptic feedback should invert the value")
        hapticEnabled.toggle()
        XCTAssertTrue(hapticEnabled, "Toggling again should restore the value")
    }

    // MARK: - Theme Mode Cycling Tests

    func testThemeMode_cycling_shouldWorkCorrectly() {
        let modes = ThemeMode.allCases
        XCTAssertEqual(modes[0], .system)
        XCTAssertEqual(modes[1], .light)
        XCTAssertEqual(modes[2], .dark)
    }

    // MARK: - AudioManager State Tests

    func testAudioManager_singleton_shouldExist() {
        let manager = AudioManager.shared
        XCTAssertNotNil(manager, "AudioManager.shared should not be nil")
    }

    func testAudioManager_singleton_shouldBeSameInstance() {
        let manager1 = AudioManager.shared
        let manager2 = AudioManager.shared
        XCTAssertTrue(manager1 === manager2, "AudioManager.shared should return the same instance")
    }

    // MARK: - HapticManager State Tests

    func testHapticManager_singleton_shouldExist() {
        let manager = HapticManager.shared
        XCTAssertNotNil(manager, "HapticManager.shared should not be nil")
    }

    func testHapticManager_singleton_shouldBeSameInstance() {
        let manager1 = HapticManager.shared
        let manager2 = HapticManager.shared
        XCTAssertTrue(manager1 === manager2, "HapticManager.shared should return the same instance")
    }

    // MARK: - SettingsManager State Tests

    func testSettingsManager_singleton_shouldExist() {
        let manager = SettingsManager.shared
        XCTAssertNotNil(manager, "SettingsManager.shared should not be nil")
    }

    func testSettingsManager_singleton_shouldBeSameInstance() {
        let manager1 = SettingsManager.shared
        let manager2 = SettingsManager.shared
        XCTAssertTrue(manager1 === manager2, "SettingsManager.shared should return the same instance")
    }

    // MARK: - Volume Tests

    func testVolume_validRange_shouldBeBetween0And1() {
        let minVolume: Float = 0.0
        let maxVolume: Float = 1.0
        let musicVolume: Float = 0.3 // Default music volume

        XCTAssertGreaterThanOrEqual(musicVolume, minVolume, "Volume should be at least 0")
        XCTAssertLessThanOrEqual(musicVolume, maxVolume, "Volume should be at most 1")
    }

    func testVolume_clamp_shouldWorkCorrectly() {
        func clampVolume(_ volume: Float) -> Float {
            return max(0, min(1, volume))
        }

        XCTAssertEqual(clampVolume(-0.5), 0.0, "Negative volume should be clamped to 0")
        XCTAssertEqual(clampVolume(0.5), 0.5, "Valid volume should remain unchanged")
        XCTAssertEqual(clampVolume(1.5), 1.0, "Volume above 1 should be clamped to 1")
    }

    // MARK: - Settings Persistence Key Tests

    func testSettingsKeys_shouldBeUnique() {
        let keys = [
            "musicEnabled",
            "soundEffectsEnabled",
            "hapticFeedbackEnabled",
            "themeMode"
        ]
        let uniqueKeys = Set(keys)
        XCTAssertEqual(keys.count, uniqueKeys.count, "All settings keys should be unique")
    }

    // MARK: - Reset Progress Logic Tests

    func testResetProgress_shouldPreserveBadges() {
        // When resetting progress, badges should be preserved
        let badgesBeforeReset = ["firstSteps", "gettingStarted"]
        let sessionsBeforeReset = 50

        // After reset, sessions should be 0 but badges remain
        let sessionsAfterReset = 0
        let badgesAfterReset = badgesBeforeReset // Badges preserved

        XCTAssertEqual(sessionsAfterReset, 0, "Sessions should be cleared after reset")
        XCTAssertEqual(badgesAfterReset, badgesBeforeReset, "Badges should be preserved after reset")
    }

    func testResetProgress_shouldClearUnlocks() {
        // When resetting progress, difficulty unlocks should be cleared
        let unlocksBeforeReset = [2, 3]
        let unlocksAfterReset: [Int] = []

        XCTAssertTrue(unlocksAfterReset.isEmpty, "Unlocks should be cleared after reset")
        XCTAssertFalse(unlocksBeforeReset.isEmpty, "Unlocks existed before reset")
    }

    func testResetProgress_shouldClearGameSessions() {
        // When resetting progress, game sessions should be cleared
        var gameSessions = ["session1", "session2", "session3"]
        gameSessions.removeAll()

        XCTAssertTrue(gameSessions.isEmpty, "Game sessions should be cleared after reset")
    }

    // MARK: - Theme Display Tests

    func testThemeMode_displayName_shouldBeHumanReadable() {
        XCTAssertEqual(ThemeMode.system.displayName, "System")
        XCTAssertEqual(ThemeMode.light.displayName, "Light")
        XCTAssertEqual(ThemeMode.dark.displayName, "Dark")
    }

    func testThemeMode_icon_shouldBeSFSymbolName() {
        // Icons should be valid SF Symbol names
        let validIconPrefixes = ["sun", "moon", "gear", "circle", "sparkle"]

        for mode in ThemeMode.allCases {
            let hasValidPrefix = validIconPrefixes.contains { mode.icon.contains($0) }
            XCTAssertTrue(hasValidPrefix || !mode.icon.isEmpty, "\(mode) icon should be a valid SF Symbol")
        }
    }

    // MARK: - Audio State Management Tests

    func testAudioState_pauseResume_shouldTrackCorrectly() {
        var isPlaying = true
        var isPaused = false

        // Pause
        isPaused = true
        isPlaying = false
        XCTAssertTrue(isPaused, "Audio should be marked as paused")
        XCTAssertFalse(isPlaying, "Audio should not be playing when paused")

        // Resume
        isPaused = false
        isPlaying = true
        XCTAssertFalse(isPaused, "Audio should not be paused after resume")
        XCTAssertTrue(isPlaying, "Audio should be playing after resume")
    }

    func testAudioState_stop_shouldResetState() {
        var isPlaying = true
        var isPaused = false

        // Stop
        isPlaying = false
        isPaused = false

        XCTAssertFalse(isPlaying, "Audio should not be playing after stop")
        XCTAssertFalse(isPaused, "Audio should not be paused after stop")
    }

    // MARK: - Settings Validation Tests

    func testSettings_allSettingsTypes_shouldBeValid() {
        // Music enabled should be boolean
        let musicEnabled: Bool = true
        XCTAssertTrue(musicEnabled is Bool, "Music enabled should be a boolean")

        // Sound effects enabled should be boolean
        let soundEnabled: Bool = true
        XCTAssertTrue(soundEnabled is Bool, "Sound effects enabled should be a boolean")

        // Haptic feedback enabled should be boolean
        let hapticEnabled: Bool = true
        XCTAssertTrue(hapticEnabled is Bool, "Haptic feedback enabled should be a boolean")

        // Theme mode should be ThemeMode enum
        let themeMode: ThemeMode = .system
        XCTAssertTrue(themeMode is ThemeMode, "Theme mode should be ThemeMode enum")
    }

    // MARK: - Haptic Feedback Type Tests

    func testHapticFeedback_types_shouldExist() {
        // Test that all haptic feedback types can be triggered without crashing
        // In a real test, these would be called on actual HapticManager
        let hapticTypes = ["buttonTap", "correctAnswer", "wrongAnswer", "gameComplete"]
        XCTAssertEqual(hapticTypes.count, 4, "There should be 4 haptic feedback types")
    }

    // MARK: - Combined Settings Tests

    func testSettings_allDisabled_shouldBeValid() {
        let musicEnabled = false
        let soundEnabled = false
        let hapticEnabled = false

        // All settings disabled should be a valid state
        XCTAssertFalse(musicEnabled)
        XCTAssertFalse(soundEnabled)
        XCTAssertFalse(hapticEnabled)
    }

    func testSettings_allEnabled_shouldBeValid() {
        let musicEnabled = true
        let soundEnabled = true
        let hapticEnabled = true

        // All settings enabled should be a valid state
        XCTAssertTrue(musicEnabled)
        XCTAssertTrue(soundEnabled)
        XCTAssertTrue(hapticEnabled)
    }

    func testSettings_mixedState_shouldBeValid() {
        // Music on, sound off, haptic on
        let musicEnabled = true
        let soundEnabled = false
        let hapticEnabled = true

        XCTAssertTrue(musicEnabled)
        XCTAssertFalse(soundEnabled)
        XCTAssertTrue(hapticEnabled)
    }
}

File: MiszIQAndroid/app/src/test/java/com/example/misziq/SettingsTests.kt
package com.example.misziq

import com.example.misziq.audio.SoundEffect
import com.example.misziq.data.preferences.ThemeMode
import org.junit.Test
import org.junit.Assert.*

/**
 * Unit tests for settings system, theme modes, and audio/haptic functionality
 */
class SettingsTests {

    // ==================== THEME MODE TESTS ====================

    @Test
    fun themeMode_system_shouldExist() {
        val themeMode = ThemeMode.SYSTEM
        assertNotNull("System theme mode should exist", themeMode)
    }

    @Test
    fun themeMode_light_shouldExist() {
        val themeMode = ThemeMode.LIGHT
        assertNotNull("Light theme mode should exist", themeMode)
    }

    @Test
    fun themeMode_dark_shouldExist() {
        val themeMode = ThemeMode.DARK
        assertNotNull("Dark theme mode should exist", themeMode)
    }

    @Test
    fun themeMode_allCases_shouldHaveDisplayName() {
        for (mode in ThemeMode.values()) {
            assertFalse("$mode should have a display name", mode.displayName.isEmpty())
        }
    }

    @Test
    fun themeMode_count_shouldBe3() {
        assertEquals("There should be exactly 3 theme modes", 3, ThemeMode.values().size)
    }

    @Test
    fun themeMode_displayNames_shouldBeCorrect() {
        assertEquals("System", ThemeMode.SYSTEM.displayName)
        assertEquals("Light", ThemeMode.LIGHT.displayName)
        assertEquals("Dark", ThemeMode.DARK.displayName)
    }

    @Test
    fun themeMode_valueOf_shouldWork() {
        assertEquals(ThemeMode.SYSTEM, ThemeMode.valueOf("SYSTEM"))
        assertEquals(ThemeMode.LIGHT, ThemeMode.valueOf("LIGHT"))
        assertEquals(ThemeMode.DARK, ThemeMode.valueOf("DARK"))
    }

    @Test(expected = IllegalArgumentException::class)
    fun themeMode_invalidValueOf_shouldThrow() {
        ThemeMode.valueOf("INVALID")
    }

    // ==================== SOUND EFFECT TESTS ====================

    @Test
    fun soundEffect_allCases_shouldHaveResourceName() {
        for (effect in SoundEffect.values()) {
            assertFalse("$effect should have a resource name", effect.resourceName.isEmpty())
        }
    }

    @Test
    fun soundEffect_count_shouldBe4() {
        assertEquals("There should be exactly 4 sound effects", 4, SoundEffect.values().size)
    }

    @Test
    fun soundEffect_resourceNames_shouldBeCorrect() {
        assertEquals("correct_chime", SoundEffect.CORRECT_ANSWER.resourceName)
        assertEquals("wrong_buzz", SoundEffect.WRONG_ANSWER.resourceName)
        assertEquals("button_tap", SoundEffect.BUTTON_TAP.resourceName)
        assertEquals("celebration", SoundEffect.GAME_COMPLETE.resourceName)
    }

    @Test
    fun soundEffect_valueOf_shouldWork() {
        assertEquals(SoundEffect.CORRECT_ANSWER, SoundEffect.valueOf("CORRECT_ANSWER"))
        assertEquals(SoundEffect.WRONG_ANSWER, SoundEffect.valueOf("WRONG_ANSWER"))
        assertEquals(SoundEffect.BUTTON_TAP, SoundEffect.valueOf("BUTTON_TAP"))
        assertEquals(SoundEffect.GAME_COMPLETE, SoundEffect.valueOf("GAME_COMPLETE"))
    }

    // ==================== SETTINGS DEFAULT VALUES TESTS ====================

    @Test
    fun settings_musicEnabled_defaultShouldBeTrue() {
        val defaultValue = true
        assertTrue("Music should be enabled by default", defaultValue)
    }

    @Test
    fun settings_soundEffectsEnabled_defaultShouldBeTrue() {
        val defaultValue = true
        assertTrue("Sound effects should be enabled by default", defaultValue)
    }

    @Test
    fun settings_hapticFeedbackEnabled_defaultShouldBeTrue() {
        val defaultValue = true
        assertTrue("Haptic feedback should be enabled by default", defaultValue)
    }

    @Test
    fun settings_themeMode_defaultShouldBeSystem() {
        val defaultMode = ThemeMode.SYSTEM
        assertEquals("Default theme mode should be system", ThemeMode.SYSTEM, defaultMode)
    }

    // ==================== SETTINGS TOGGLE TESTS ====================

    @Test
    fun settings_toggleMusic_shouldInvertValue() {
        var musicEnabled = true
        musicEnabled = !musicEnabled
        assertFalse("Toggling music should invert the value", musicEnabled)
        musicEnabled = !musicEnabled
        assertTrue("Toggling again should restore the value", musicEnabled)
    }

    @Test
    fun settings_toggleSoundEffects_shouldInvertValue() {
        var soundEnabled = true
        soundEnabled = !soundEnabled
        assertFalse("Toggling sound effects should invert the value", soundEnabled)
        soundEnabled = !soundEnabled
        assertTrue("Toggling again should restore the value", soundEnabled)
    }

    @Test
    fun settings_toggleHapticFeedback_shouldInvertValue() {
        var hapticEnabled = true
        hapticEnabled = !hapticEnabled
        assertFalse("Toggling haptic feedback should invert the value", hapticEnabled)
        hapticEnabled = !hapticEnabled
        assertTrue("Toggling again should restore the value", hapticEnabled)
    }

    // ==================== THEME MODE CYCLING TESTS ====================

    @Test
    fun themeMode_cycling_shouldWorkCorrectly() {
        val modes = ThemeMode.values()
        assertEquals(ThemeMode.SYSTEM, modes[0])
        assertEquals(ThemeMode.LIGHT, modes[1])
        assertEquals(ThemeMode.DARK, modes[2])
    }

    @Test
    fun themeMode_ordinal_shouldBeCorrect() {
        assertEquals(0, ThemeMode.SYSTEM.ordinal)
        assertEquals(1, ThemeMode.LIGHT.ordinal)
        assertEquals(2, ThemeMode.DARK.ordinal)
    }

    // ==================== VOLUME TESTS ====================

    @Test
    fun volume_validRange_shouldBeBetween0And1() {
        val minVolume = 0.0f
        val maxVolume = 1.0f
        val musicVolume = 0.3f // Default music volume

        assertTrue("Volume should be at least 0", musicVolume >= minVolume)
        assertTrue("Volume should be at most 1", musicVolume <= maxVolume)
    }

    @Test
    fun volume_clamp_shouldWorkCorrectly() {
        fun clampVolume(volume: Float): Float {
            return maxOf(0f, minOf(1f, volume))
        }

        assertEquals("Negative volume should be clamped to 0", 0.0f, clampVolume(-0.5f), 0.001f)
        assertEquals("Valid volume should remain unchanged", 0.5f, clampVolume(0.5f), 0.001f)
        assertEquals("Volume above 1 should be clamped to 1", 1.0f, clampVolume(1.5f), 0.001f)
    }

    // ==================== SETTINGS PERSISTENCE KEY TESTS ====================

    @Test
    fun settingsKeys_shouldBeUnique() {
        val keys = listOf(
            "music_enabled",
            "sound_effects_enabled",
            "haptic_feedback_enabled",
            "theme_mode"
        )
        val uniqueKeys = keys.toSet()
        assertEquals("All settings keys should be unique", keys.size, uniqueKeys.size)
    }

    // ==================== RESET PROGRESS LOGIC TESTS ====================

    @Test
    fun resetProgress_shouldPreserveBadges() {
        val badgesBeforeReset = listOf("firstSteps", "gettingStarted")
        val sessionsBeforeReset = 50

        // After reset, sessions should be 0 but badges remain
        val sessionsAfterReset = 0
        val badgesAfterReset = badgesBeforeReset // Badges preserved

        assertEquals("Sessions should be cleared after reset", 0, sessionsAfterReset)
        assertEquals("Badges should be preserved after reset", badgesBeforeReset, badgesAfterReset)
    }

    @Test
    fun resetProgress_shouldClearUnlocks() {
        val unlocksBeforeReset = listOf(2, 3)
        val unlocksAfterReset = emptyList<Int>()

        assertTrue("Unlocks should be cleared after reset", unlocksAfterReset.isEmpty())
        assertFalse("Unlocks existed before reset", unlocksBeforeReset.isEmpty())
    }

    @Test
    fun resetProgress_shouldClearGameSessions() {
        val gameSessions = mutableListOf("session1", "session2", "session3")
        gameSessions.clear()

        assertTrue("Game sessions should be cleared after reset", gameSessions.isEmpty())
    }

    // ==================== AUDIO STATE MANAGEMENT TESTS ====================

    @Test
    fun audioState_pauseResume_shouldTrackCorrectly() {
        var isPlaying = true
        var isPaused = false

        // Pause
        isPaused = true
        isPlaying = false
        assertTrue("Audio should be marked as paused", isPaused)
        assertFalse("Audio should not be playing when paused", isPlaying)

        // Resume
        isPaused = false
        isPlaying = true
        assertFalse("Audio should not be paused after resume", isPaused)
        assertTrue("Audio should be playing after resume", isPlaying)
    }

    @Test
    fun audioState_stop_shouldResetState() {
        var isPlaying = true
        var isPaused = false

        // Stop
        isPlaying = false
        isPaused = false

        assertFalse("Audio should not be playing after stop", isPlaying)
        assertFalse("Audio should not be paused after stop", isPaused)
    }

    // ==================== HAPTIC FEEDBACK TYPE TESTS ====================

    @Test
    fun hapticFeedback_types_shouldExist() {
        val hapticTypes = listOf("buttonTap", "correctAnswer", "wrongAnswer", "gameComplete")
        assertEquals("There should be 4 haptic feedback types", 4, hapticTypes.size)
    }

    // ==================== COMBINED SETTINGS TESTS ====================

    @Test
    fun settings_allDisabled_shouldBeValid() {
        val musicEnabled = false
        val soundEnabled = false
        val hapticEnabled = false

        assertFalse(musicEnabled)
        assertFalse(soundEnabled)
        assertFalse(hapticEnabled)
    }

    @Test
    fun settings_allEnabled_shouldBeValid() {
        val musicEnabled = true
        val soundEnabled = true
        val hapticEnabled = true

        assertTrue(musicEnabled)
        assertTrue(soundEnabled)
        assertTrue(hapticEnabled)
    }

    @Test
    fun settings_mixedState_shouldBeValid() {
        // Music on, sound off, haptic on
        val musicEnabled = true
        val soundEnabled = false
        val hapticEnabled = true

        assertTrue(musicEnabled)
        assertFalse(soundEnabled)
        assertTrue(hapticEnabled)
    }

    // ==================== DATASTORE KEY TESTS ====================

    @Test
    fun dataStoreKeys_musicEnabled_shouldBeCorrect() {
        val key = "music_enabled"
        assertEquals("music_enabled", key)
    }

    @Test
    fun dataStoreKeys_soundEffectsEnabled_shouldBeCorrect() {
        val key = "sound_effects_enabled"
        assertEquals("sound_effects_enabled", key)
    }

    @Test
    fun dataStoreKeys_hapticFeedbackEnabled_shouldBeCorrect() {
        val key = "haptic_feedback_enabled"
        assertEquals("haptic_feedback_enabled", key)
    }

    @Test
    fun dataStoreKeys_themeMode_shouldBeCorrect() {
        val key = "theme_mode"
        assertEquals("theme_mode", key)
    }

    // ==================== THEME MODE CONVERSION TESTS ====================

    @Test
    fun themeMode_fromString_shouldWork() {
        fun themeModeFromString(value: String): ThemeMode {
            return try {
                ThemeMode.valueOf(value.uppercase())
            } catch (e: IllegalArgumentException) {
                ThemeMode.SYSTEM
            }
        }

        assertEquals(ThemeMode.SYSTEM, themeModeFromString("SYSTEM"))
        assertEquals(ThemeMode.LIGHT, themeModeFromString("light"))
        assertEquals(ThemeMode.DARK, themeModeFromString("DARK"))
        assertEquals(ThemeMode.SYSTEM, themeModeFromString("invalid")) // Falls back to SYSTEM
    }

    @Test
    fun themeMode_toString_shouldWork() {
        assertEquals("SYSTEM", ThemeMode.SYSTEM.name)
        assertEquals("LIGHT", ThemeMode.LIGHT.name)
        assertEquals("DARK", ThemeMode.DARK.name)
    }

    // ==================== SOUND EFFECT ENUMERATION TESTS ====================

    @Test
    fun soundEffect_allValues_shouldBeEnumerable() {
        val effects = SoundEffect.values().toList()
        assertTrue("Should be able to enumerate all sound effects", effects.isNotEmpty())
        assertEquals(4, effects.size)
    }

    @Test
    fun soundEffect_ordinals_shouldBeSequential() {
        assertEquals(0, SoundEffect.CORRECT_ANSWER.ordinal)
        assertEquals(1, SoundEffect.WRONG_ANSWER.ordinal)
        assertEquals(2, SoundEffect.BUTTON_TAP.ordinal)
        assertEquals(3, SoundEffect.GAME_COMPLETE.ordinal)
    }

    // ==================== VIBRATION EFFECT TESTS ====================

    @Test
    fun vibrationEffect_validDuration_shouldBePositive() {
        val shortVibration = 50L // milliseconds
        val mediumVibration = 100L
        val longVibration = 200L

        assertTrue("Short vibration should be positive", shortVibration > 0)
        assertTrue("Medium vibration should be positive", mediumVibration > 0)
        assertTrue("Long vibration should be positive", longVibration > 0)
    }

    @Test
    fun vibrationEffect_durations_shouldBeReasonable() {
        val maxReasonableDuration = 1000L // 1 second max for haptic feedback
        val buttonTapDuration = 50L
        val successDuration = 100L
        val errorDuration = 150L
        val completeDuration = 200L

        assertTrue("Button tap vibration should be under 1 second", buttonTapDuration < maxReasonableDuration)
        assertTrue("Success vibration should be under 1 second", successDuration < maxReasonableDuration)
        assertTrue("Error vibration should be under 1 second", errorDuration < maxReasonableDuration)
        assertTrue("Complete vibration should be under 1 second", completeDuration < maxReasonableDuration)
    }

    // ==================== SETTINGS STATE CONSISTENCY TESTS ====================

    @Test
    fun settings_shouldMaintainConsistency() {
        // Simulate setting changes
        var musicEnabled = true
        var soundEnabled = true
        var hapticEnabled = true
        var themeMode = ThemeMode.SYSTEM

        // Change each setting
        musicEnabled = false
        soundEnabled = false
        hapticEnabled = false
        themeMode = ThemeMode.DARK

        // Verify all changes persisted
        assertFalse(musicEnabled)
        assertFalse(soundEnabled)
        assertFalse(hapticEnabled)
        assertEquals(ThemeMode.DARK, themeMode)
    }

    @Test
    fun settings_independentToggle_shouldNotAffectOthers() {
        var musicEnabled = true
        var soundEnabled = true
        var hapticEnabled = true

        // Toggle only music
        musicEnabled = false

        // Other settings should remain unchanged
        assertFalse("Music should be disabled", musicEnabled)
        assertTrue("Sound should still be enabled", soundEnabled)
        assertTrue("Haptic should still be enabled", hapticEnabled)
    }
}
